lib/pure/decimal.nim

```diff
@@ -0,0 +1,1040 @@
+    if newValue == noValue:
+      return
+    let currentPlaces = -exponent
+    if currentPlaces != newValue:
+      let digits = generateDigits(number, upper)
+      var diff = (currentPlaces - newValue).int16
+      if diff > 0: # take away precision
+        let temp = digits.toTempSignificandArray()
+        let (newDigits, roundingEffect) = bankersRoundingToEven(temp, diff)
+        number = createDecimal(number.negative, newDigits.toTempSignificandArray, exponent + roundingEffect)
+      else:
+        let significance = digitCount(digits)
+        let newSig = significance - diff
+        if newSig > significandSize:
+          raise newException(
+            ValueError, 
+            "Too many decimal places ($1 places for $2). A 128-bit decimal can only hold 34 digits ($3 rose to $4)."
+              .format(newValue, $number, significance, newSig)
+            )
+        else:
+          let newDigits = shiftDecimalsLeftWithZero(digits, -diff)
+          number = createDecimal(number.negative, newDigits.toTempSignificandArray, exponent + diff)
+  of dkInfinite:
+    discard
+  of dkNaN:
+    discard
+
+
+#
+# ops
+#
+
+proc `~==`(left: Decimal, right: Decimal): bool =
+  ## Determines whether two numbers match in the context of their
+  ## least-common significance.
+  ## 
+  ## For example:
+  ## 
+  ##   assert (1.1'm  ~== 1.0'm ) == false
+  ##   assert (1.0'm  ~== 1.0'm ) == true
+  ##   assert (1.0'm  ~== 1.01'm) == true
+  ##   assert (1.01'm ~== 1.0'm ) == true
+  ##   assert (1.01'm ~== 1.00'm) == false
+  ## 
+  ## Use this with care with currency or financial work:
+  ## 
+  ##   assert (5'm    ~== 5.23'm) == true
+  ##   assert (5.00'm ~== 5.23'm) == false 
+  # let (leftKind, leftExponent, leftUpper) = determineKindExponentAndUpper(left)
+  # let (rightKind, rightExponent, rightUpper) = determineKindExponentAndUpper(right)
+  # case leftKind:
+  # of dkValued:
+  #   if rightKind != dkValues:
+  #     result = false
+  #     return
+  #   if (leftExponent > rightExponent): # these values are -(places)
+  #     let stripped = right.strip(-leftExponent)
+  #     result = left == stripped
+  #   else:
+  #     let stripped = left.strip(-rightExponent)
+  #     result = stripped == right
+  # of dkInfinite:
+  #   result = left == right
+  # of dkNaN:
+  #   result = left == right
+  result = false # TODO
+
+# proc '<'
+
+# proc round  # bankers rounding
+
+# proc roundUp
+
+# proc strip
+
+#
+# assignment / input
+#
+
+proc newDecimal*(numberString: string, places: int = noValue): Decimal =
+  var (decimalKind, negativeFlag, significand, exponent) = parseFromString(numberString)
+  case decimalKind:
+  of dkNaN:
+    result = nan
+  of dkInfinite:
+    if negativeFlag:
+      result = negativeInfinity
+    else:
+      result = infinity
+  of dkValued:
+    result = createDecimal(negativeFlag, significand, exponent)
+    result.places = places
+
+proc `'m`*(numberString: string, places: int = noValue): Decimal =
```

Comment:
What's the reasoning for using `'m`, instead of something more descriptive like `'dec` or `'decimal`?

Reply:
It's a suffix used by two other languages. C# and a different one that doesn't come to mind right now.

 (Most languages don't support suffixes. Even fewer support user-defined ones.)

Philosophically, I suspect they chose `m` since `d` would be confused for "double", 'e' would be confused with "exponent", 'l' with "long", etc.

Most programs will not used `decimals`, but when they do the suffix will be used a ***lot***. So, having a single letter makes things less cluttered.

Reply:
Yeah, the 'm' is fine.